packages/api/src/middlewares/permissions/index.ts

```diff
@@ -0,0 +1,85 @@
+import { NextFunction, Request, Response } from 'express';
+import {
+  PermissionAccess,
+  PermissionRoles,
+  Predicate,
+  Workspace,
+} from '@src/models';
+import { IAuthRequest } from '@middlewares/auth/types';
+import {
+  ErrorTypes,
+  NotFound,
+  Unauthorized,
+  UnauthorizedErrorTitles,
+} from '@utils/error';
+
+export interface IPredicatePermissionMiddlewareOptions {
+  predicateSelector: (req: Request) => string;
+  permissions: PermissionRoles[];
+}
+
+export const predicatePermissionMiddleware = (
+  options: IPredicatePermissionMiddlewareOptions,
+) => {
+  return async (req: Request, _: Response, next: NextFunction) => {
+    try {
+      const { user, workspace: authWorkspace }: IAuthRequest = req;
+
+      const predicateId = options.predicateSelector(req);
+      const predicate = await Predicate.findOne({
+        where: { id: predicateId },
+        relations: ['workspace'],
+      });
+
+      // Require predicate to be found
+      if (!predicate) {
+        throw new NotFound({
+          type: ErrorTypes.NotFound,
+          title: 'Predicate not found',
+          detail: `Predicate with id ${predicateId} not found`,
+        });
+      }
+
+      const { workspace: predicateWorkspace } = predicate;
+      const isSameWorkspace = predicateWorkspace.id === authWorkspace.id;
+      const isMember = await Workspace.createQueryBuilder('workspace')
+        .innerJoin('workspace.members', 'members')
+        .where('members.id = :userId', { userId: user.id })
+        .andWhere('workspace.id = :workspaceId', {
+          workspaceId: predicateWorkspace.id,
+        })
+        .getExists();
```

Comment:
I think it doesn't work when the user is a member of the predicate but not a member of the workspace.